cloud/validations.go

```diff
@@ -0,0 +1,279 @@
+// describing clouds, including regions, supported auth types etc.
+
+package cloud
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/gojsonschema"
+	"gopkg.in/yaml.v2"
+)
+
+//ValidationWarning are JSON schema validation errors used to warn users about
+//potential schema violations
+type ValidationWarning struct {
+	Messages []string
+}
+
+func (e *ValidationWarning) Error() string {
+	str := ""
+	for _, msg := range e.Messages {
+		str = fmt.Sprintf("%s\n%s", str, msg)
+	}
+
+	return str
+}
+
+var cloudSetSchema = map[string]interface{}{
+	"type": "object",
+	"properties": map[string]interface{}{
+		"clouds": map[string]interface{}{
+			"type":                 "object",
+			"additionalProperties": cloudSchema,
+		},
+	},
+	"additionalProperties": false,
+}
+
+var cloudSchema = map[string]interface{}{
+	"type": "object",
+	"properties": map[string]interface{}{
+		"name":        map[string]interface{}{"type": "string"},
+		"type":        map[string]interface{}{"type": "string"},
+		"description": map[string]interface{}{"type": "string"},
+		"auth-types": map[string]interface{}{
+			"type":  "array",
+			"items": map[string]interface{}{"type": "string"},
+		},
+		"endpoint":          map[string]interface{}{"type": "string"},
+		"identity-endpoint": map[string]interface{}{"type": "string"},
+		"storage-endpoint":  map[string]interface{}{"type": "string"},
+		"config":            map[string]interface{}{"type": "object"},
+		"regions":           regionsSchema,
+		"region-config":     map[string]interface{}{"type": "object"},
+	},
+	"additionalProperties": false,
+}
+
+var regionsSchema = map[string]interface{}{
+	"type": "object",
+	"additionalProperties": map[string]interface{}{
+		"type": "object",
+		"properties": map[string]interface{}{
+			"endpoint":          map[string]interface{}{"type": "string"},
+			"identity-endpoint": map[string]interface{}{"type": "string"},
+			"storage-endpoint":  map[string]interface{}{"type": "string"},
+		},
+		"additionalProperties": false,
+	},
+}
+
+// ValidateCloudSet reports any erroneous properties found in cloud metadata
+// YAML. If there are no erroneous properties, then ValidateCloudSet returns nil
+// otherwise it return an error listing all erroneous properties and possible
+// suggestion.
+func ValidateCloudSet(data []byte) error {
+	return validateCloud(data, &cloudSetSchema)
+}
+
+// ValidateOneCloud is like ValidateCloudSet but validates the metadata for only
+// one cloud and not multiple.
+func ValidateOneCloud(data []byte) error {
+	return validateCloud(data, &cloudSchema)
+}
+
+func validateCloud(data []byte, jsonSchema *map[string]interface{}) error {
+	var body interface{}
+	if err := yaml.Unmarshal(data, &body); err != nil {
+		return errors.Annotate(err, "cannot unmarshal yaml cloud metadata")
+	}
+
+	jsonBody := yamlToJSON(body)
+	invalidKeys, err := validateCloudMetaData(jsonBody, jsonSchema)
+
```

Comment:
nit: please keep the error check directly underneath where the error is produced. they're tightly related, so belong together